#159987708 Like and dislike functionality

iamuchejude opened the pull request:
### What does this PR do?
Add like and dislike feature for articles

### Description of the task?
* Create endpoints for like and dislike functionality
* Create endpoint to get rating/reactions for an article
* Write relevant test for like and dislike feature

### How to manually test
* Clone the repository
* cd to `thor-ah`
* Setup postgres database
* Create `.env` using the format in `.env.sample`
* Run `npm install`
* Run `npm run start:dev`
* Test the following endpoints using postman or any other available alternative software
**POST**  /api/articles/:article_slug/reactions
**GET** /api/articles/:article_slug/reactions 

### What are the relevant Pivotal Tracker Story?
[#159987708](https://www.pivotaltracker.com/story/show/159987708)

iamuchejude committed:
Setting up model for like and dislike

iamuchejude committed:
Setting up model for like and dislike

iamuchejude committed:
feat(likeDislike): implement like and dislike functionality

- write tests for like and dislike enpoints
- create like and dislike endpoints

[Delivers #159987708]

iamuchejude committed:
feat(likeDislike): implement like and dislike functionality

- write tests for like and dislike enpoints
- create like and dislike endpoints

[Delivers #159987708]

iamuchejude committed:
feat(likeDislike): fix houndci issue

emekafredy commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+              });
+            }
+            likesDislikes.forEach((likeAndDislike) => {
+              if (likeAndDislike.status === 'likedßß') {
```
On line 152, shouldn't it be 'liked' instead of 'likedBB'?

emekafredy commented on server/models/likedislike.js:
```diff
@@ -0,0 +1,20 @@
+module.exports = (sequelize, DataTypes) => {
```
You have two identical models for 'likedislike'. Same applies to migrations

iamuchejude committed:
feat(likeDislike): implement fix in test and controller

waleCloud commented:
@iamuchejude 

looking through your commit messages, i can see upto 4 commits that has `New Update` as the commit message attached to it.

I will suggest you be more descriptive about your messages, take a look at @d-beloved 's commit from the PR he raised.
I couldn't help but to point this out while looking through the codebase.

CEOehis reviewed (approved):
LGTM

emekafredy reviewed (changes requested):
Sorry, I think you have conflicts to resolve in 'server/controllers/users.js'

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+          }
+        })
+          .then((likeOrDislike) => {
+            if (likeOrDislike === null || likeOrDislike === undefined) {
```
if (!likeOrDislike) {

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+    if (!reaction || reaction.trim().length < 1) {
+      return res.status(400).json({
+        status: 'error',
+        message: 'Please provide a valid reaction .',
```
message: 'No reaction provided.',

iamuchejude committed:
feat(likeDislike): implement fix

- resolve conflict in usersController
- refractor likesDislikesController

emekafredy reviewed (approved):
LGTM

iamuchejude committed:
feat(likeDislike): implement fix in GET likeDislike endpoint

d-beloved commented on server/migrations/20180907145330-create-likes-dislikes.js:
```diff
@@ -0,0 +1,31 @@
+      allowNull: false,
+    },
+    status: {
+      type: Sequelize.ENUM('liked', 'disliked'),
```
you can change the ENUM value to `1` and `0`, so when an article is liked, it takes 1 and when it is disliked, it takes `0`

iamuchejude replied:
Thanks for that. I have implemented the feedback

lauragift21 reviewed (changes requested):
For naming your models, you can add underscore between the words making it likes_dislikes.js.

iamuchejude committed:
feat(likeDislike): refractor like and dislike endpoints

- rename status column in likesDislikes table to reaction
- change enumeration values for reaction from 'liked and disliked' to '1 and 0'
- ensure tests pass with new changes

iamuchejude committed:
feat(likeDislike): fix GET likeDislike response error

lauragift21 commented:
LGTM

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+ */
+class likesDislikesController {
+  /**
+   * @desc Like an article
```
misleading comment

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+   * @return {obj} returns an object
+   */
+  static likeOrDislike(req, res, next) {
+    const { articleSlug } = req.params;
```
use article id as the check

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+          message: "Reaction must either be 'like' or 'dislike'",
+        });
+      }
+    }
```
duplicate check. just keep the second one. i.e. 
if ((reaction === 'like') !== true) {
      if ((reaction === 'dislike') !== true) {
        return res.status(400).json({
          status: 'error',
          message: "Reaction must either be 'like' or 'dislike'",
        });
      }
    }

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+
+    Article.findOne({
+      where: {
+        slug: articleSlug,
```
use articleID

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+              })
+                .then(() => res.status(200).json({
+                  status: 'success',
+                  message: `Article ${reaction}d successfully.`,
```
nice

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+      return res.status(400).json({
+        status: 'error',
+        message: 'Please provide a valid article slug.',
+      });
```
No valid article ID provided

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+          },
+        })
+          .then((likesDislikes) => {
+            if (likesDislikes === null) {
```
if (!likesDislikes) {

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+            message: 'Article was not found.',
+          });
+        }
+        LikesDislikes.findAll({
```
let likes,
let dislikes
LikesDislikes.findAll({
where: {
            articleId: article.id,
            reaction: 0}).then((allLikes) => {if (!allLikes} { likes === 0} else {likes = allLikes.Length})


same for dislikes. return data: {likes}, {dislikes}
          },

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,167 @@
+  }
+}
+
+export default likesDislikesController;
```
check if a user liked or disliked an article so it can be used on the front end

iamuchejude committed:
feat(likeDislike): implement fix on like and dislike feature

- refractor like and dislike endpoints
- add a new endpoint to check if user has reacted to an article
- fix broken test issues

iamuchejude committed:
feat(likeDislike): implement fix on like and dislike feature

- refractor like and dislike endpoints
- add a new endpoint to check if user has reacted to an article
- fix broken test issues

iamuchejude committed:
feat(likeDislike): refractor likesDislikesController

- remove duplicate check

emekafredy reviewed (changes requested):
On line 24 of your likes-dislikes controller, message should be a property of an error object, as we have agreed. something like this:
error: {
   message: 'error message here'
}

Same applies for lines 31, 39, 54, 117, 130, 173 and 185.

iamuchejude committed:
feat(likeDislike): refractor like and dislike endpoints and tests

- update error response format in likeDislike controller
- update tests for like and dislike endpoints to ensure they pass
- resolve conflicts
- updated jsdoc comment

emekafredy reviewed (commented):
LGTM

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,227 @@
+
+  /**
+   * @desc Get all reaction status of an article for a particular user
+   * @param {obj} req request body
```
misleading comment

iamuchejude replied:
Okay...I will update it

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,227 @@
+import db from '../models';
+
```
import { Article, LikesDislikes } from '../models';

aknwosu commented on server/controllers/likeDislike.js:
```diff
@@ -0,0 +1,227 @@
+import db from '../models';
```
import { Article, LikesDislikes } from '../models';

iamuchejude replied:
Thanks. I will do the needful
